### PR TITLE
Add infrastructure for refactors

### PR DESCRIPTION
**Pull request by zhengbli**
This PR adds the infrastructure for refactors.

The PR adds one new kind of code fixes as well as refactors:
1. "diagnostic-generating code fixes": a special kind of code fixes that would generate non-error diagnostics. After normal syntactic check and semantic check, the language service will do an extra pass of the AST to get the non-error diagnostics generated by registered "diagnostic-generating code fixes". Then when the user interacts with the new diagnostics in code, a query for corresponding code actions will be triggered, just like with other code fixes.

2. "refactors": to get a refactor it is a two-step procedure. First, the editor will ask about all "applicable refactor info" at a given location / range, the returned "applicable refactor info" would only contain basic information like refactor names, it wouldn't compute the code actions yet. Then after the user chose a particular refactor, then the editor will send another request to compute the corresponding code actions.

### CODE REVIEW

File: src/harness/unittests/tsserverProjectSystem.ts [hunk 1]
```diff
@@ -3446,10 +3450,18 @@ namespace ts.projectSystem {
                 assert.equal(e1.event, "syntaxDiag");
                 host.clearOutput();
 
+                // the semanticDiag message
                 host.runQueuedImmediateCallbacks();
-                assert.equal(host.getOutput().length, 2, "expect 2 messages");
+                assert.equal(host.getOutput().length, 1, "expect 1 messages");
```
Comment (aozgaa) [inline src/harness/unittests/tsserverProjectSystem.ts]: "messages" should be "message"

File: src/server/client.ts [hunk 2]
```diff
@@ -692,6 +692,50 @@ namespace ts.server {
             return response.body.map(entry => this.convertCodeActions(entry, fileName));
         }
 
+        getRefactorDiagnostics(fileName: string, range?: TextRange): RefactorDiagnostic[] {
+            const startLineOffset = this.positionToOneBasedLineOffset(fileName, range.pos);
```
Comment (aozgaa) [inline src/server/client.ts]: Does this crash if `range === undefined` ?

File: src/server/protocol.ts [hunk 3]
```diff
@@ -395,17 +398,63 @@ namespace ts.server.protocol {
...
     export interface CodeFixRequest extends Request {
         command: CommandTypes.GetCodeFixes;
         arguments: CodeFixRequestArgs;
     }
 
-    /**
-      * Instances of this interface specify errorcodes on a specific location in a sourcefile.
-      */
-    export interface CodeFixRequestArgs extends FileRequestArgs {
+    export interface TextRangeRequestArgs extends FileRequestArgs {
```
Comment (aozgaa) [inline src/server/protocol.ts]: Can you name the first two members `line` and `offset` to more closely coincide with `FileLocationRequestArgs` and`Location`? Or use a `Location` for the start and end?

File: src/server/session.ts [hunk 4]
```diff
@@ -1411,15 +1417,72 @@ namespace ts.server {
             return ts.getSupportedCodeFixes();
         }
 
+        private extractStartAndEndPositionFromTextRangeRequestArgs(args: protocol.TextRangeRequestArgs, scriptInfo: ScriptInfo): { startPosition: number, endPosition: number } {
+            const startPosition = getStartPosition();
+            const endPosition = getEndPosition();
+            return { startPosition, endPosition };
+
+            function getStartPosition() {
+                return args.startPosition !== undefined ? args.startPosition : scriptInfo.lineOffsetToPosition(args.startLine, args.startOffset);
```
Comment (aozgaa) [inline src/server/session.ts]: Do you want to save the result in `args.startPosition` (resp. `args.endPosition`)?

File: src/services/refactorProvider.ts [hunk 5]
```diff
@@ -0,0 +1,92 @@
+﻿/* @internal */
+namespace ts {
+    interface BaseRefactor {
+        /** Description of the refactor to display in the UI of the editor */
```
Comment (aozgaa) [inline src/services/refactorProvider.ts]: should this have a property `canBeSuggested: boolean;` ?

File: src/services/refactorProvider.ts [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+        canBeSuggested: false;
+
+        /** A fast syntactic check to see if the refactor is applicable at given position. */
+        isApplicableForRange(range: TextRange, context: LightRefactorContext): boolean;
+    }
+
+    export type Refactor = SuggestableRefactor | NonSuggestableRefactor;
+
+    export interface LightRefactorContext {
+        nonBoundSourceFile: SourceFile;
```
Comment (aozgaa) [inline src/services/refactorProvider.ts]: Can you add a comment explaining what `bound` and `nonBound` means?

File: src/services/refactorProvider.ts [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+            for (const code in nonSuggestableRefactors) {
+                const refactor = nonSuggestableRefactors[code];
+                if (refactor && refactor.isApplicableForRange(range, context)) {
+                    results.push(createRefactorDiagnostic(refactor, range));
+                }
+            }
+            return results;
+        }
+
+        export function getSuggestedRefactorDiagnosticsForNode(node: Node, context: RefactorContext) {
```
Comment (aozgaa) [inline src/services/refactorProvider.ts]: Would it be right to name this `getSuggestable...`? That would be more consistent with the naming of the interfaces.

File: src/services/services.ts [hunk 8]
```diff
@@ -1900,11 +1901,56 @@ namespace ts {
...
+                const diags = refactor.getSuggestedRefactorDiagnosticsForNode(node, context);
+                if (diags.length > 0) {
+                    addRange(result, diags);
+                }
+
+                forEachChild(node, visitor);
+            }
+        }
+
+        function getCodeActionsForRefactorAtPosition(
```
Comment (aozgaa) [inline src/services/services.ts]: Do we have a convention for when we split function parametersacross multiple lines?

File: src/server/client.ts (hunk 2 above)
Comment (zhengbli) [inline src/server/client.ts]: Here the `range` is not supposed to be nullable, as we only use it to test the session API `GetRefactorsForRange`. Will update

File: src/server/protocol.ts (hunk 3 above)
Comment (zhengbli) [inline src/server/protocol.ts]: Maybe `FileRangeRequestArgs` would fit better. But using `location` for start and end would be an unnecessary breaking change for editors already consuming this type.

File: src/services/services.ts (hunk 8 above)
Comment (zhengbli) [inline src/services/services.ts]: I don't think so, normally i do that when I have to use the horizontal scroll bar on my laptop if putting all in one line.

File: src/server/session.ts [hunk 9]
```diff
@@ -1429,15 +1435,63 @@ namespace ts.server {
...
+            let position: number;
+            let textRange: TextRange;
+            if (this.isLocation(args)) {
+                position = getPosition(args);
+            }
+            else {
+                const { startPosition, endPosition } = this.getStartAndEndPosition(args, scriptInfo);
+                textRange = { pos: startPosition, end: endPosition };
+            }
+            return { position, textRange };
```
Comment (RyanCavanaugh) [inline src/server/session.ts]: position may be undefined, is this intended? Initialize at line 1450 if so

File: src/server/session.ts [hunk 10]
```diff
@@ -1450,14 +1504,16 @@ namespace ts.server {
...
 
-            function getStartPosition() {
-                return args.startPosition !== undefined ? args.startPosition : scriptInfo.lineOffsetToPosition(args.startLine, args.startOffset);
-            }
-
-            function getEndPosition() {
-                return args.endPosition !== undefined ? args.endPosition : scriptInfo.lineOffsetToPosition(args.endLine, args.endOffset);
-            }
+        private getStartAndEndPosition(args: protocol.FileRangeRequestArgs, scriptInfo: ScriptInfo) {
+            const startPosition = args.startPosition !== undefined
```
Comment (RyanCavanaugh) [inline src/server/session.ts]: This is really confusing, can we just use a normal if here

File: src/services/codeFixProvider.ts [hunk 11]
```diff
@@ -18,22 +26,38 @@ namespace ts {
...
         }
 
         export function getSupportedErrorCodes() {
             return Object.keys(codeFixes);
         }
 
+        export function getCodeFixDiagnosticsForNode(context: CodeFixDiagnoseContext, node: Node): Diagnostic[] | undefined {
+            let result: Diagnostic[];
+            for (const codeFix of diagnosticGeneratingCodeFixes) {
+                const newDiag = codeFix.createCodeFixDiagnosticIfApplicable(node, context);
```
Comment (RyanCavanaugh) [inline src/services/codeFixProvider.ts]: This member is declared as optional, but isn't checked for `undefined` in this method

File: src/services/refactorProvider.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        description: string;
+
+        /** Compute the associated code actions */
+        getCodeActions(context: RefactorContext, positionOrRange: number | TextRange): CodeAction[];
+
+        /** A fast syntactic check to see if the refactor is applicable at given position. */
+        isApplicableForPositionOrRange(context: LightRefactorContext, positionOrRange: number | TextRange): boolean;
+    }
+
+    export interface LightRefactorContext {
```
Comment (RyanCavanaugh) [inline src/services/refactorProvider.ts]: Comment defining what "Light" means

File: src/services/refactorProvider.ts [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+            let results: ApplicableRefactorInfo[];
+            refactors.forEach(refactor => {
+                if (refactor.isApplicableForPositionOrRange(context, positionOrRange)) {
+                    (results || (results = [])).push({ refactorName: refactor.name, description: refactor.description });
+                }
+            });
+            return results;
+        }
+
+        export function getRefactorCodeActions(
```
Comment (RyanCavanaugh) [inline src/services/refactorProvider.ts]: Why does this return an empty array vs `getApplicableRefactors` returns `undefined` if the array would have been empty?

**Comment by RyanCavanaugh** [comment]
Should we add a trivial refactor with this just so we can have some testcases?

File: src/services/codeFixProvider.ts (hunk 11 above)
Comment (zhengbli) [inline src/services/codeFixProvider.ts]: in line 41 it already checked the existence of `createCodeFixDiagnosticIfApplicable`

Comment (RyanCavanaugh) [inline src/services/codeFixProvider.ts]: :+1:

**Comment by zhengbli** [comment]
Sure, I'm adding one now will update soon

**Comment by zhengbli** [comment]
I added two things:
1. a diagnostic-generating code fix: converting functions that has prototype assignment to ES6 classes
2. a sample refactor: add `Async` suffix to async function declarations

**Comment by RyanCavanaugh** [comment]
@mhegazy any other comments?

**Comment by mhegazy** [comment]
closing in favor of https://github.com/Microsoft/TypeScript/pull/15569

**Pull request closed without merging**